File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public int pairSum(ListNode head) {
        
        // STEP 1:- find the mid

        ListNode slow = head ;
        ListNode fast = head ;

        while(fast != null){
            slow = slow.next ;
            fast = fast.next.next ;
        }


        //STEP 2:- reverse form the mid part
        ListNode head2 = reverse(slow) ;

        // STEP 3 :- final check to fins the max sum
        int sum = 0 ;

        while(head2 != null){
            int currSum = head.val + head2.val ;
            sum = Math.max(currSum, sum) ;

            head2 = head2.next ;
            head = head.next ;
        }

        return sum ;
    }

    private ListNode reverse(ListNode head) {
        ListNode prev = null ;
        ListNode curr ;

        while(head != null){
            curr = head ;
            head = head.next ;
            curr.next = prev ;
            prev = curr ;
        }

        return prev ;
    }
}

File: 2236-maximum-twin-sum-of-a-linked-list/maximum-twin-sum-of-a-linked-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func pairSum(head *ListNode) int {
    
    slow := head
    fast := head


    // 1. Finding Mid
    for fast != nil{
        slow = slow.Next
        fast = fast.Next.Next
    }

    // 2. Reverse the Other half

    head2 := reverse(slow) 

    // 3. Claculate Sum

    max := math.MinInt32 

    for head2 != nil{
        curr := head.Val + head2.Val

       if max < curr {

          max = curr
       }

       head = head.Next
       head2 = head2.Next
    }


    return max 

}


func reverse(head *ListNode) *ListNode{
    var prev *ListNode

    for head != nil{
        curr := head
        head = head.Next 
        curr.Next = prev 
        prev = curr
    }

    return prev
}